src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/StopApprenticeship/StopApprenticeshipCommandHandler.cs

```diff
@@ -0,0 +1,110 @@
+using SFA.DAS.CommitmentsV2.Messages.Commands;
+using SFA.DAS.CommitmentsV2.Messages.Events;
+using SFA.DAS.CommitmentsV2.Shared.Interfaces;
+using SFA.DAS.CommitmentsV2.Types;
+using SFA.DAS.Encoding;
+using SFA.DAS.NServiceBus.Services;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.CommitmentsV2.Application.Commands.StopApprenticeship
+{
+    public class StopApprenticeshipCommandHandler : AsyncRequestHandler<StopApprenticeshipCommand>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _dbContext;
+        private readonly ICurrentDateTime _currentDate;
+        private readonly IAuthenticationService _authenticationService;
+        private readonly IEventPublisher _eventPublisher;
+        private readonly IMessageSession _nserviceBusContext;
+        private readonly IEncodingService _encodingService;
+        private readonly ILogger<StopApprenticeshipCommandHandler> _logger;
+        private const string StopNotificationEmailTemplate = "ProviderApprenticeshipStopNotification";
+
+        public StopApprenticeshipCommandHandler(
+            Lazy<ProviderCommitmentsDbContext> dbContext,
+            ICurrentDateTime currentDate,
+            IEventPublisher eventPublisher,
+            IAuthenticationService authenticationService,
+            IMessageSession nserviceBusContext,
+            IEncodingService encodingService,
+            ILogger<StopApprenticeshipCommandHandler> logger)
+        {
+            _dbContext = dbContext;
+            _currentDate = currentDate;
+            _eventPublisher = eventPublisher;
+            _authenticationService = authenticationService;
+            _nserviceBusContext = nserviceBusContext;
+            _encodingService = encodingService;
+            _logger = logger;
+        }
+
+        protected async override Task Handle(StopApprenticeshipCommand command, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _logger.LogInformation($"Begin stopping apprenticeShip. Apprenticeship-Id:{command.ApprenticeshipId}");
+
+                var party = _authenticationService.GetUserParty();
+                CheckPartyIsValid(party);
+
+                var apprenticeship = await _dbContext.Value.Apprenticeships
+                    .Include(t => t.DataLockStatus)
+                    .Include(s => s.Cohort).ThenInclude(c => c.AccountLegalEntity)
+                    .SingleOrDefaultAsync(x => x.Id == command.ApprenticeshipId);
+
+                apprenticeship.ValidateApprenticeshipForStop(command.StopDate, command.AccountId, _currentDate);
+
+                apprenticeship.StopApprenticeship(command.StopDate, command.MadeRedundant, command.UserInfo, party);
+
+                _logger.LogInformation($"Stopped apprenticeShip. Apprenticeship-Id:{command.ApprenticeshipId}");
+
+                await _eventPublisher.Publish(new ApprenticeshipStoppedEvent
```

Comment:
This event should be emitted from with the `StopApprenticeship` method itself. The apprenticeship class has access to a `Publish` method for this purpose. (See other state changes as an example).

Reply:
Have moved the event inside the `Apprenticeship` object.